src/QMCWaveFunctions/LCAO/SoaAtomicBasisSet.h

```diff
@@ -736,6 +736,220 @@ class SoaAtomicBasisSet
+    }
+
+    // Batched radial evaluation
+    {
+      ScopedTimer local(rnl_timer_);
+      MultiRnl.batched_evaluateVGL_multiCenter(r, rnl_vgl, Rmax, num_centers);
+    }
+
+    // Batched angular evaluation
+    {
+      ScopedTimer local(ylm_timer_);
+      Ylm.batched_evaluateVGL_multiCenter(dr, ylm_vgl, num_centers);
+    }
+
+    {
+      ScopedTimer local_timer(psi_timer_);
+      const int bset_size = BasisSetSize;
+      auto* phase_fac_ptr = periodic_image_phase_factors_.device_data();
+      auto* LM_ptr        = LM.device_data();
+      auto* NL_ptr        = NL.device_data();
+
+      // Device pointers for radial and angular components
+      RealType* restrict phi_ptr   = rnl_vgl.device_data_at(0, 0, 0, 0);
+      RealType* restrict dphi_ptr  = rnl_vgl.device_data_at(1, 0, 0, 0);
+      RealType* restrict d2phi_ptr = rnl_vgl.device_data_at(2, 0, 0, 0);
+
+      const RealType* restrict ylm_v_ptr = ylm_vgl.device_data_at(0, 0, 0, 0);
+      const RealType* restrict ylm_x_ptr = ylm_vgl.device_data_at(1, 0, 0, 0);
+      const RealType* restrict ylm_y_ptr = ylm_vgl.device_data_at(2, 0, 0, 0);
+      const RealType* restrict ylm_z_ptr = ylm_vgl.device_data_at(3, 0, 0, 0);
+      const RealType* restrict ylm_l_ptr = ylm_vgl.device_data_at(4, 0, 0, 0);
+
+      ///Assumed to be already on DEVICE
+      //V,Gx,Gy,Gz,L
+      auto* restrict psi_ptr    = psi_vgl.device_data_at(0, 0, 0);
+      auto* restrict dpsi_x_ptr = psi_vgl.device_data_at(1, 0, 0);
+      auto* restrict dpsi_y_ptr = psi_vgl.device_data_at(2, 0, 0);
+      auto* restrict dpsi_z_ptr = psi_vgl.device_data_at(3, 0, 0);
+      auto* restrict d2psi_ptr  = psi_vgl.device_data_at(4, 0, 0);
+
+      // Get multi-center data pointers
+      auto* correctphase_ptr = atom_bs_leader.mw_mem_handle_.getResource().correctphase.device_data();
+      auto* r_ptr            = r.device_data();
+      auto* dr_ptr           = dr.device_data();
+
+      PRAGMA_OFFLOAD("omp target teams distribute parallel for collapse(3) \
+                  is_device_ptr(ylm_v_ptr, ylm_x_ptr, ylm_y_ptr, ylm_z_ptr, ylm_l_ptr, \
+                  phi_ptr, dphi_ptr, d2phi_ptr,psi_ptr, dpsi_x_ptr, dpsi_y_ptr, dpsi_z_ptr, d2psi_ptr) \
+                  is_device_ptr(phase_fac_ptr, LM_ptr, NL_ptr, correctphase_ptr, r_ptr, dr_ptr,basis_offsets_ptr)")
+      for (size_t i_c = 0; i_c < num_centers; ++i_c)
+      {
+        for (int i_e = 0; i_e < nElec; ++i_e)
+        {
+          for (int ib = 0; ib < bset_size; ++ib)
+          {
+            const int nl = NL_ptr[ib];
+            const int lm = LM_ptr[ib];
+            VT psi = 0, dpsi_x = 0, dpsi_y = 0, dpsi_z = 0, d2psi = 0;
+
+            // Get center-specific offset
+            const size_t center_offset = basis_offsets_ptr[i_c];
+
+            for (int i_xyz = 0; i_xyz < Nxyz; ++i_xyz)
+            {
+              // Calculate global index accounting for center batching
+              const size_t global_idx = i_xyz + Nxyz * (i_e + i_c * nElec);
```

Comment:
the ordering here looks like it is assuming a data layout as [num_centers, nElec, Nxyz], but in the array dims above it looked like [nElec, Nxyx * num_centers]

Reply:
ok, I took a closer look, and it looks like at least the ordering for dr/Ylm and r/Rnl are consistent, so that's why the results are correct, but it still seems to be inconsistent with the way the dimensions are described elsewhere in the code (e.g. when resizing at the top of this function).

Reply:
you could change the ordering of the loops and the index labels to match the array dimensions above, or you could change the array dimensions, but then they wouldn't agree with how they're described in the multiwalker mem

Reply:
For optimal performance, the fastest index should match the inner most loop under `parallel for` for maxing out vector performance. However, your inner most loop is `ib` which doesn't affect global_idx and thus, I don't think it matters much. We can defer kernel tuning in separate steps. Just make sure the code is consistent with documentation in the source code.

Reply:
For now it passes CI tests...